File: 89_Tic-Tac-Toe/java/src/TicTacToe2.java
import java.util.Scanner;
import java.util.Random;

/**
 * @author Ollie Hensman-Crook
 */
public class TicTacToe2 {
    public static void main(String[] args) {
        Board gameBoard = new Board();
        Random compChoice = new Random();
        char yourChar;
        char compChar;
        Scanner in = new Scanner(System.in);

        System.out.println("              TIC-TAC-TOE");
        System.out.println("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println("\nTHE BOARD IS NUMBERED: ");
        System.out.println(" 1  2  3\n 4  5  6\n 7  8  9\n");

        while (true) {
            // ask if the player wants to be X or O and if their input is valid set their
            // play piece as such
            System.out.println("DO YOU WANT 'X' OR 'O'");
            while (true) {
                try {
                    char input;
                    input = in.next().charAt(0);

                    if (input == 'X' || input == 'x') {
                        yourChar = 'X';
                        compChar = 'O';
                        break;
                    } else if (input == 'O' || input == 'o') {
                        yourChar = 'O';
                        compChar = 'X';
                    } else {
                        System.out.println("THATS NOT 'X' OR 'O', TRY AGAIN");
                        in.nextLine();
                    }

                } catch (Exception e) {
                    System.out.println("THATS NOT 'X' OR 'O', TRY AGAIN");
                    in.nextLine();
                }
            }

            while (true) {
                System.out.println("WHERE DO YOU MOVE");

                // check the user can move where they want to and if so move them there
                while (true) {
                    int input;
                    try {
                        input = in.nextInt();
                        if (gameBoard.getBoardValue(input) == ' ') {
                            gameBoard.setArr(input, yourChar);
                            break;
                        } else {
                            System.out.println("INVALID INPUT, TRY AGAIN");
                        }
                        in.nextLine();
                    } catch (Exception e) {
                        System.out.println("INVALID INPUT, TRY AGAIN");
                        in.nextLine();
                    }
                }

                gameBoard.printBoard();
                System.out.println("THE COMPUTER MOVES TO");

                while (true) {
                    int position = 1 + compChoice.nextInt(9);
                    if (gameBoard.getBoardValue(position) == ' ') {
                        gameBoard.setArr(position, compChar);
                        break;
                    }
                }

                gameBoard.printBoard();

                // if there is a win print if player won or the computer won and ask if they
                // want to play again
                if (gameBoard.checkWin(yourChar)) {
                    System.out.println("YOU WIN, PLAY AGAIN? (Y/N)");
                    gameBoard.clear();
                    while (true) {
                        try {
                            char input;
                            input = in.next().charAt(0);

                            if (input == 'Y' || input == 'y') {
                                break;
                            } else if (input == 'N' || input == 'n') {
                                System.exit(0);
                            } else {
                                System.out.println("THATS NOT 'Y' OR 'N', TRY AGAIN");
                                in.nextLine();
                            }

                        } catch (Exception e) {
                            System.out.println("THATS NOT 'Y' OR 'N', TRY AGAIN");
                            in.nextLine();
                        }
                    }
                    break;
                } else if (gameBoard.checkWin(compChar)) {
                    System.out.println("YOU LOSE, PLAY AGAIN? (Y/N)");
                    gameBoard.clear();
                    while (true) {
                        try {
                            char input;
                            input = in.next().charAt(0);

                            if (input == 'Y' || input == 'y') {
                                break;
                            } else if (input == 'N' || input == 'n') {
                                System.exit(0);
                            } else {
                                System.out.println("THATS NOT 'Y' OR 'N', TRY AGAIN");
                                in.nextLine();
                            }

                        } catch (Exception e) {
                            System.out.println("THATS NOT 'Y' OR 'N', TRY AGAIN");
                            in.nextLine();
                        }
                    }
                    break;
                } else if (gameBoard.checkDraw()) {
                    System.out.println("DRAW, PLAY AGAIN? (Y/N)");
                    gameBoard.clear();
                    while (true) {
                        try {
                            char input;
                            input = in.next().charAt(0);

                            if (input == 'Y' || input == 'y') {
                                break;
                            } else if (input == 'N' || input == 'n') {
                                System.exit(0);
                            } else {
                                System.out.println("THATS NOT 'Y' OR 'N', TRY AGAIN");
                                in.nextLine();
                            }

                        } catch (Exception e) {
                            System.out.println("THATS NOT 'Y' OR 'N', TRY AGAIN");
                            in.nextLine();
                        }
                    }
                    break;
                }

            }
        }
    }
}

File: 89_Tic-Tac-Toe/kotlin/TicTacToe2.kt
import java.util.Random
import kotlin.system.exitProcess

/**
 * @author John Long based on Java from Ollie Hensman-Crook
 */
private val compChoice = Random()
private val gameBoard = Board()

fun main() {
    println("              TIC-TAC-TOE")
    println("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    println("\nTHE BOARD IS NUMBERED: ")
    println(" 1  2  3\n 4  5  6\n 7  8  9\n")
    while (true) {
        // Let the player choose whether to be X or O (Player.X or Player.O)
        val (human, computer) = readXOrO()
        while (true) {
            // Get a valid move from the user and then move there
            val validMoveIndex = readValidMove()
            gameBoard.setArr(validMoveIndex, human)
            gameBoard.printBoard()

            // Computer randomly fills a square (if the game isn't already over)
            // This uses Kotlin's null handling and will only set the board
            // if validRandomMove returned a non-null value
            validRandomMove()?.let {
                gameBoard.setArr(it, computer)
                gameBoard.printBoard()
            }

            // if there is a win print if player won or the computer won and ask if they
            // want to play again
            when {
                gameBoard.isWinFor(human) -> {
                    checkPlayAgain("YOU WIN")
                    break
                }
                gameBoard.isWinFor(computer) -> {
                    checkPlayAgain("YOU LOSE")
                    break
                }
                gameBoard.isDraw() -> {
                    checkPlayAgain("DRAW")
                    break
                }
            }
        }
    }
}

private fun checkPlayAgain(result: String) {
    println("$result, PLAY AGAIN? (Y/N)")
    gameBoard.clear()
    if (!readYesOrNo()) exitProcess(0)
}

private fun readYesOrNo(): Boolean {
    while (true) {
        when (readLine()?.get(0)?.uppercaseChar()) {
            'Y' -> return true
            'N' -> return false
            else -> println("THAT'S NOT 'Y' OR 'N', TRY AGAIN")
        }
    }
}

private fun validRandomMove(): Int? {
    if (gameBoard.isDraw() || gameBoard.isWinFor(Player.O) || gameBoard.isWinFor(Player.X)) return null
    println("THE COMPUTER MOVES TO")
    // keep generating a random value until we find one that is null (unset)
    return generateSequence { 1 + compChoice.nextInt(9) }.first { gameBoard.getBoardValue(it) == null }
}

private fun readValidMove(): Int {
    println("WHERE DO YOU MOVE")
    while (true) {
        val input = readln().toIntOrNull()
        if (input != null && gameBoard.getBoardValue(input) == null) {
            return input
        } else {
            println("INVALID INPUT, TRY AGAIN")
        }
    }
}

private fun readXOrO(): Pair<Player, Player> {
    println("DO YOU WANT 'X' OR 'O'")
    while (true) {
        when (readln()[0].uppercaseChar()) {
            'X' -> return Player.X to Player.O
            'O' -> return Player.O to Player.X
            else -> println("THAT'S NOT 'X' OR 'O', TRY AGAIN")
        }
    }
}
